### PR TITLE
Adding a transport action to get cluster formation info

### PR DESCRIPTION
**Pull request by masseyke**
This PR exposes ClusterFormationFailureHelper.ClusterFormationState in a transport action (ClusterFormationInfoAction), with the intention that it be used as part of the master stability health check. This was started as https://github.com/elastic/elasticsearch/pull/85782, but we realized that we needed more fine-grained information than just the string description of the reason the cluster formation failed.
2022-04-18 Coordination health
This action will be used in branch 1.2.2.4 of the attached diagram (near the bottom right corner). That is, when we detect that there has been no master node in the last 30 seconds (default), and when PeerFinder does not believe that there is an elected master node, and when the current node is master eligible, we will use this transport action to asynchronously send a request to all master-eligible nodes to gather their ClusterFormationStates.

### CODE REVIEW

**Comment by masseyke** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java [hunk 1]
```diff
@@ -114,17 +120,120 @@ public String toString() {
...
+                in.readLong(),
+                new VotingConfiguration(in),
+                new VotingConfiguration(in),
+                in.readImmutableList(TransportAddress::new),
+                in.readImmutableList(DiscoveryNode::new),
+                in.readLong(),
+                in.readBoolean(),
+                readStatusInfo(in),
+                in.readList(
+                    in1 -> new JoinStatus(
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java:216]: could we have a more meaningful name for the inner reader?

File: server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java [hunk 2]
```diff
@@ -114,17 +120,120 @@ public String toString() {
...
+                        new DiscoveryNode(in1),
+                        in1.readLong(),
+                        in1.readString(),
+                        new TimeValue(in1.readLong(), TimeUnit.valueOf(in1.readString()))
+                    )
+                )
+            );
+        }
+
+        private static StatusInfo readStatusInfo(StreamInput in) throws IOException {
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java:226]: Would it be worth making `StatusInfo` `Writeable`?

File: server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        @Override
+        public int hashCode() {
+            return Objects.hash(clusterFormationState);
+        }
+    }
+
+    /**
+     * This transport action fetches the ClusterFormationState from a remote node.
+     */
+    public static class TransportAction extends HandledTransportAction<
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java:120]: See @DaveCTurner 's comment on the previous PR

https://github.com/elastic/elasticsearch/pull/85782/files#r849933849

File: server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    public static final ClusterFormationInfoAction INSTANCE = new ClusterFormationInfoAction();
+    public static final String NAME = "cluster:internal/formation/info";
+
+    private ClusterFormationInfoAction() {
+        super(NAME, ClusterFormationInfoAction.Response::new);
+    }
+
+    public static class Request extends ActionRequest {
+
+        public Request() {}
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java:44]: We'll have to request this information from all master eligible nodes so the request should be able to express that?

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 5]
```diff
@@ -284,7 +284,7 @@ public Coordinator(
         this.nodeHealthService = nodeHealthService;
     }
 
-    private ClusterFormationState getClusterFormationState() {
+    public ClusterFormationState getClusterFormationState() {
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:287]: @DaveCTurner does it make sense to send this over if the discovery cluster formation timeout has not lapsed yet?

File: server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java [hunk 6]
```diff
@@ -114,17 +120,120 @@ public String toString() {
...
+                    clusterState.version(),
+                    clusterState.getLastCommittedConfiguration(),
+                    clusterState.getLastAcceptedConfiguration()
+                ),
+                statusInfo,
+                inFlightJoinStatuses
+            );
+        }
+
+        private static boolean calculateHasDiscoveredQuorum(
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java:178]: nit: would it make sense to obey the `isXXX`/`hasXXX` naming convention for methods returning boolean?

**Review by andreidan — CHANGES_REQUESTED** [review]
Thanks for working on this Keith

Left a few comments

File: server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java:120]: This is good tho, it's just a simple `HandledTransportAction` here.

File: server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java (hunk 1 above)
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java]: JoinStatus is now Writeable and readable from the wire.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java (hunk 4 above)
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java:44]: Okido, sounds good

File: server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java (hunk 3 above)
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java:120]: Ah of course - I just realised it does the needful 
```
        transportService.registerRequestHandler(actionName, executor, false, canTripCircuitBreaker, requestReader, new TransportHandler());
```

Sorry for the noise Keith

File: server/src/main/java/org/elasticsearch/monitor/StatusInfo.java [hunk 7]
```diff
@@ -35,6 +37,20 @@ public Status getStatus() {
...
+        if (statusString == null) {
+            return null;
+        }
+        return Status.valueOf(statusString);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalString(status == null ? null : status.name());
+        out.writeOptionalString(info);
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/monitor/StatusInfo.java:51]: Are we changing the semantics for `StatusInfo` ? I don't think these fields are optional now. I think we should keep them non-null (ie. mandatory).

**Review by andreidan — APPROVED** [review]
LGTM, thanks for working on this Keith

File: server/src/main/java/org/elasticsearch/monitor/StatusInfo.java (hunk 7 above)
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/monitor/StatusInfo.java:51]: There was nothing preventing them from being null. I thought it was very unlikely that someone would pass in nulls (what would be the point?). But if they do, I don't want it to blow up during serialization. From what I can tell there's no good way to enforce null checks on a record. What do you think?

Comment (andreidan) [reply server/src/main/java/org/elasticsearch/monitor/StatusInfo.java:51]: I caught up with @dakrone on this and we were thinking of aligning the object to the requirements (ie. non-null fields) via `Objects.requireNonNull` calls in the constructor, alongside strict (non-optional) serialisation.

What do you think @masseyke ?

Comment (masseyke) [reply server/src/main/java/org/elasticsearch/monitor/StatusInfo.java:51]: I'm all for enforcing non-nullable fields. I'll change the serialization to assume no nulls, and we can enforce it on a subsequent PR.

**Pull request merged**